[items] Refactor `metadata` & `itemchannllink` APIs

florian-h05 opened the pull request:
Fixes #189.

## Description

This PR completely refactors the `metadata` and the `itemchannellink` APIs.
Their old versions have been removed (since metadata was unable to support configuration, I'm not sure how many even used it).

The new `metadata` API is available under the `items` namespace and is able to add and change metadata configuration as well.
`itemchannellink` got a few internal improvements and a renaming.

## To Do

- [x] Update CHANGELOG

## Testing

The following script tests all methods of these APIs:

```javascript
// replaceMetadata and replaceItemChannelLink are used inside the replaceItem method
items.replaceItem({
  type: 'String',
  name: 'Hallway_Light',
  label: 'Hallway Light',
  category: 'light',
  groups: ['Hallway', 'Light'],
  tags: ['Lightbulb'],
  channels: {
    'binding:thing:device:hallway#light': {},
    'binding:thing:device:livingroom#light': { 
      profile: 'system:follow' 
    }
  },
  metadata: {
    expire: '10m,command=1',
    stateDescription: {
      config: {
        pattern: '%d%%',
        options: '1=Red, 2=Green, 3=Blue'
      }
    }
  }
});

// getMetadata
console.log('Metadata expire = ' + JSON.stringify(items.getItem('Hallway_Light').getMetadata('expire')));
console.log('Metadata stateDescription configuration = ' + items.getItem('Hallway_Light').getMetadata('stateDescription').configuration);

// removeMetadata
console.warn('Removing metadata expire ...')
items.getItem('Hallway_Light').removeMetadata('expire');
console.log('Metadata expire = ' + items.getItem('Hallway_Light').getMetadata('expire'));

// getItemChannelLink
console.log(items.metadata.itemchannellink.getItemChannelLink('Hallway_Light', 'binding:thing:device:livingroom#light'))
console.log(items.metadata.itemchannellink.getItemChannelLink('Hallway_Light', 'binding:thing:device:livingroom#light').configuration)


// removeItemChannelLink
items.metadata.itemchannellink.removeItemChannelLink('Hallway_Light', 'binding:thing:device:livingroom#light');
console.log(items.metadata.itemchannellink.getItemChannelLink('Hallway_Light', 'binding:thing:device:livingroom#light'))
```

florian-h05 commented:
@rkoshak Ping.

digitaldan commented on items/items.js:
```diff
@@ -501,20 +491,11 @@ const removeItem = function (itemOrItemName) {
-      throw e;
-    }
+  const item = getItem(itemConfig.name, true);
+  if (item !== null) {
```
So it would be good to know if the method actually replaced the item or not.   Maybe the method should at least return the old item or null if it did not exist so that there is some way of knowing if it replaced or inserted?   This is how a java Map `replace()` works actually.

jpg0 replied:
FYI I used `replaceItem` by just sticking it into some scripts which needed items to be defined (typically synthetic ones) and it would ensure it was present (or completely fail with an exception) without needing any other error handling. It was more of a 'configuration-as-code' pattern however, which I know you are moving away from.

digitaldan commented on items/metadata/itemchannellink.js:
```diff
@@ -0,0 +1,139 @@
+  const existing = getItemChannelLink(itemName, channelUID);
+  if (existing === null) {
+    _addItemChannelLink(itemName, channelUID, conf);
+    return true;
```
I think returning the previous channel link or null would be better than a boolean ( i try to avoid such patterns).  Again, Java Maps ad other objects tend to do the same thing (previous value or null);

digitaldan reviewed (approved):
> Their old versions have been removed (since metadata was unable to support configuration, I'm not sure how many even used it).

So i poked around a little on this, i think the few that were using item links,  used the native Java proxy objects and did not use this library, so i think its probably a very small population if any . 

I had only a few small nits, other wise LGTM

florian-h05 commented:
@digitaldan I have addressed your review, can you please have a look at the changes?

> i think its probably a very small population if any .

I thing the same, but I'll label it as breaking to make sure that no one is suprised

digitaldan commented on items/items.js:
```diff
@@ -496,25 +487,20 @@ const removeItem = function (itemOrItemName) {
-    } else {
-      throw e;
-    }
+  const item = getItem(itemConfig.name, true);
+  if (item !== null) { // Item already existed
+    removeItem(itemConfig.name);
+    addItem(itemConfig);
+    return item;
+  } else { // Item did not exist
+    addItem(itemConfig);
+    return null;
   }
```
```suggestion
  const item = getItem(itemConfig.name, true);
  if (item !== null) { // Item already existed
    removeItem(itemConfig.name);
  }
  addItem(itemConfig);
  return item;
```

digitaldan commented on items/metadata/itemchannellink.js:
```diff
@@ -0,0 +1,139 @@
+ */
+const replaceItemChannelLink = function (itemName, channelUID, conf) {
+  const existing = getItemChannelLink(itemName, channelUID);
+  if (existing === null) {
+    _addItemChannelLink(itemName, channelUID, conf);
+    return null;
+  } else {
+    _updateItemChannelLink(itemName, channelUID, conf);
+    return existing;
+  }
```
```suggestion
  if (existing === null) {
    _addItemChannelLink(itemName, channelUID, conf);
  } else {
    _updateItemChannelLink(itemName, channelUID, conf);
  }
  return existing;
```

digitaldan reviewed (approved):
I had two more code style changes recommendations.   Feel free to merge either way.  Thanks!

rkoshak commented:
Sorry I missed the ping. It all looks as I would expect from the metadata side (I never messed with channel links in rules so don't have much to say there.

proggprod commented:
Been playing around with this. Would be nice if I could remove all channel links or metadata by excluding the name or some other wildcard. Sometimes I don't want to specify the name or channel link, I just want to remove/replace.

florian-h05 commented:
IIRC, openHAB Core provides that possibility, so I could add this to the library as well. Can you please open an issue and describe your feature request?
